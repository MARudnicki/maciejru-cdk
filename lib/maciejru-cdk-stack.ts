// import * as sqs from 'aws-cdk-lib/aws-sqs';

import {Construct, Stack, StackProps} from "@aws-cdk/core";

export class MaciejruCdkStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here

    // example resource
    // const queue = new sqs.Queue(this, 'MaciejruCdkQueue', {
    //   visibilityTimeout: cdk.Duration.seconds(300)
    // });
  }
}
