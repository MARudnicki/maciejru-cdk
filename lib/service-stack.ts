import {Construct, Stack, StackProps} from "@aws-cdk/core";
import {CfnParametersCode, Code, Function, Runtime} from "@aws-cdk/aws-lambda";
import {HttpApi, HttpMethod} from "@aws-cdk/aws-apigatewayv2";
import {HttpProxyIntegration, LambdaProxyIntegration} from "@aws-cdk/aws-apigatewayv2-integrations";

export class ServiceStack extends Stack {

    public readonly serviceCode: CfnParametersCode

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        this.serviceCode = Code.fromCfnParameters()

        const lambda: Function = new Function(this, "ServiceLambda", {
            runtime: Runtime.NODEJS_14_X,
            handler: 'src/lambda.handler',
            code: this.serviceCode,
            functionName: 'ServiceLambda'
        })

        const innerApiGateway: HttpApi = new HttpApi(this, 'InnerApiGateway', {
            defaultIntegration: new LambdaProxyIntegration({
                handler: lambda
            }),
            apiName: 'InnerApiGateway'
        })

        const outerApiGateway = new HttpApi(this, 'outerApiGateway')
        outerApiGateway.addRoutes({
            path: "/",
            methods: [HttpMethod.GET],
            integration: new HttpProxyIntegration({
                url: innerApiGateway.url || "",
            })
        })

        outerApiGateway.addRoutes({
            path: "/test",
            methods: [HttpMethod.GET],
            integration: new HttpProxyIntegration({
                url: "https://google.com"
            })
        })

    }
}
