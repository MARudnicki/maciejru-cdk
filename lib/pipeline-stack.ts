import {Construct, SecretValue, Stack, StackProps} from "@aws-cdk/core";
import {Artifact, IStage, Pipeline} from "@aws-cdk/aws-codepipeline";
import {
    CloudFormationCreateUpdateStackAction,
    CodeBuildAction,
    GitHubSourceAction
} from "@aws-cdk/aws-codepipeline-actions";
import {BuildSpec, LinuxBuildImage, PipelineProject} from "@aws-cdk/aws-codebuild";
import {ServiceStack} from "./service-stack";
import {BillingStack} from "./biling-stack";

export class PipelineStack extends Stack {

    private readonly pipeline: Pipeline;
    private readonly cdkBuildOutput: Artifact;
    private readonly lambdaBuildOutput: Artifact;


    constructor(scope: Construct, id: string, props: StackProps) {
        super(scope, id, props);

        this.pipeline = new Pipeline(this, 'Pipeline', {
            pipelineName: 'MyPipeline',
            crossAccountKeys: false,
            restartExecutionOnUpdate: true
        })

        const cdkSourceOutput = new Artifact('CdkSourceOutput')
        const lambdaSourceOutput = new Artifact('LambdaSourceOutput')

        this.pipeline.addStage({
            stageName: 'MySource',
            actions: [
                new GitHubSourceAction({
                    owner: 'MARudnicki',
                    repo: 'maciejru-cdk',
                    branch: 'master',
                    actionName: 'Pipeline_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: cdkSourceOutput
                }),
                new GitHubSourceAction({
                    owner: 'MARudnicki',
                    repo: 'express-lambda',
                    branch: 'master',
                    actionName: 'Lambda_Source',
                    oauthToken: SecretValue.secretsManager('github-token'),
                    output: lambdaSourceOutput
                })

            ]
        })

        this.cdkBuildOutput = new Artifact("CdkBuildOutput")
        this.lambdaBuildOutput = new Artifact("LambdaBuildOutput")

        this.pipeline.addStage({

            stageName: 'MyBuild',
            actions: [
                new CodeBuildAction({
                    actionName: 'CDK_Build',
                    input: cdkSourceOutput,
                    outputs: [this.cdkBuildOutput],
                    project: new PipelineProject(this, 'CdkBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-spec/cdk-build-spec.yml')
                    })
                }),
                new CodeBuildAction({
                    actionName: 'Service_Build',
                    input: lambdaSourceOutput,
                    outputs: [this.lambdaBuildOutput],
                    project: new PipelineProject(this, 'LambdaBuildProject', {
                        environment: {
                            buildImage: LinuxBuildImage.STANDARD_5_0
                        },
                        buildSpec: BuildSpec.fromSourceFilename('build-spec/lambda-build-spec.yml')
                    })
                })
            ]

        })

        this.pipeline.addStage({
            stageName: 'MyPipelineUpdate',
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: 'Pipeline_Update',
                    stackName: "MaciejruCdkStack",
                    templatePath: this.cdkBuildOutput.atPath("MaciejruCdkStack.template.json"),
                    adminPermissions: true
                })
            ]
        })
    }

    public addServiceStage(serviceStack: ServiceStack, stageName: string): IStage {
        return this.pipeline.addStage({
            stageName: stageName,
            actions: [
                new CloudFormationCreateUpdateStackAction({
                    actionName: 'Service_Update',
                    stackName: serviceStack.stackName,
                    templatePath: this.cdkBuildOutput.atPath(`${serviceStack.stackName}.template.json`),
                    adminPermissions: true,
                    parameterOverrides: {
                        ...serviceStack.serviceCode.assign(
                            this.lambdaBuildOutput.s3Location
                        )
                    },
                    extraInputs: [this.lambdaBuildOutput]
                })
            ]
        })
    }

    public addBillingStackToStage(bilingStack: BillingStack, stage: IStage) {
        stage.addAction(new CloudFormationCreateUpdateStackAction({
            actionName: 'Biling_Update',
            stackName: bilingStack.stackName,
            templatePath: this.cdkBuildOutput.atPath(
                `{bilingStack.stackName}.template.json`
            ),
            adminPermissions: true
        }))
    }
}
