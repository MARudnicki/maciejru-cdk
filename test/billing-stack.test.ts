import {BillingStack} from "../lib/biling-stack";
import {SynthUtils} from "@aws-cdk/assert";
import {App} from "@aws-cdk/core";

test('Billing Stack', ()=>{
    const app = new App()
    const stack = new BillingStack(app, "BillingStack", {
        budgetAmount: 2,
        emailAddress: "test@example.com"
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
})


