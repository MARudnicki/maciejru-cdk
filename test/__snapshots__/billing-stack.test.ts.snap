// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Billing Stack 1`] = `
Object {
  "Resources": Object {
    "BudgetED13AE41": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 2,
            "Unit": "USD",
          },
          "BudgetName": "Monthly Budget",
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 99,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": "test@example.com",
                "SubscriptionType": "EMAIL",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
  },
}
`;
